### PR TITLE
update globbing and tests to use regex for the match finding

### PR DESCRIPTION
**Pull request by bendichter**
## Motivation

@CodyCBakerPhD take a look at this and let me know what you think. This approach changes things around quite a bit to rely more on Python's regular expressions library. The approach here is to take the f-string notation and transform it into a regex expression, and then proceed from there using regex. I think moving into regex would have some advantages, as it would technically allow any user that knows regex to perform more advanced pattern matches.

## Checklist

- [ ] Have you thoroughly read our [Developer Guide](https://neuroconv.readthedocs.io/en/main/developer_guide.html)?
- [ ] Have you ensured the PR description clearly describes the problem and solutions?
- [ ] Have you checked to ensure that there aren't other open or previously closed [Pull Requests](https://github.com/catalystneuro/neuroconv/pulls) for the same change?
- [ ] If this PR fixes an issue, is the first line of the PR description `fix #XXX` where `XXX` is the issue number?
- [ ] Did you update the [CHANGLOG.md](https://github.com/catalystneuro/neuroconv/blob/main/CHANGELOG.md) file on your branch to describe your changes?

### CODE REVIEW

**Comment by bendichter** [comment]
Example usage:

```python
>>> fstring_pattern = "sub-{subject_id}/sub-{subject_id}/ses-{session_id}"
>>> pattern = generate_regex_from_fstring(fstring_pattern)
>>> print(pattern)
sub-(?P<subject_id>.+)/sub-(?P=subject_id)/ses-(?P<session_id>.+)

>>> import re
>>> re.match(pattern, "sub-001/sub-001/ses-abc")
{'subject_id': '001', 'session_id': 'abc'}
```

**Comment by bendichter** [comment]
Translation of the regex expression sub-(?P<subject_id>.+)/sub-(?P=subject_id)/ses-(?P<session_id>.+)


* `(?P<subject_id>.+)`: Match the pattern `.+` and assign it to the group name "subject_id".
* `(?P=subject_id)` find another match for "subject_id : with the same pattern rule as the first instance.
* `(?P<session_id>.+)`: Match the pattern `.+` and assign it to the group name "session_id"

**Comment by bendichter** [comment]
tests are failing due to escape character issue I have not yet figured out

**Comment by CodyCBakerPhD** [comment]
Basic approach seems sound

In the YAML implementation, people will be able to specify either an f-string pattern or a precise regex pattern (if they know it)?

I'm not expecting the majority of users to know regex, but yes if they do its good to have the option for greater customization

**Comment by bendichter** [comment]
@CodyCBakerPhD yeah, exactly. Would you be able to help with debugging on Windows? I don't have a Windows machine handy. It should just be a matter of properly handling backslashes.

**Comment by CodyCBakerPhD** [comment]
> Would you be able to help with debugging on Windows? I don't have a Windows machine handy. It should just be a matter of properly handling backslashes.

Yeah I remember something like that popping up somewhere before

File: src/neuroconv/utils/globbing.py [hunk 1]
```diff
@@ -1,90 +1,71 @@
...
-        # variable_names = ["x", "y", "z"]
-        # separators = ["a/", "b", "/c"", ""]
+
+    >>> fstring_pattern = "sub-{subject_id}/sub-{subject_id}/ses-{session_id}"
+    >>> pattern = generate_regex_from_fstring(fstring_pattern)
+    >>> print(pattern)
+    sub-(?P<subject_id>.+)/sub-(?P=subject_id)/ses-(?P<session_id>.+)
+
+    >>> import re
+    >>> re.match(pattern, "sub-001/sub-001/ses-abc")
```
Comment (bendichter) [inline src/neuroconv/utils/globbing.py:30]:
```suggestion
    >>> re.match(pattern, "sub-001/sub-001/ses-abc").groupdict()
```

**Comment by bendichter** [comment]
@CodyCBakerPhD actually, don't bother. I am working on a more robust solution now.

**Comment by bendichter** [comment]
It might be better to use the parse library for this

**Comment by h-mayorquin** [comment]
> It might be better to use the parse library for this

I remember that me and @CodyCBakerPhD discussed parse when implementing this. I remember that he pointed out some edge case that was not working in case you decide to be that way maybe the previous discussion is useful:

https://github.com/catalystneuro/nwb-conversion-tools/pull/457

**Comment by bendichter** [comment]
Thanks for the pointer, @h-mayorquin !

**Comment by bendichter** [comment]
I've been looking into this. At first, I was in favor of switching to `parse` because I thought it could parse arbitrary date formats. Then when I tried it, it was more limited than I realized. I opened an issue here: https://github.com/r1chardj0n3s/parse/issues/150

If they respond I'd like to consider reopening the issue of using parse.

**Pull request closed without merging**